### DISCUSSION TITLE
弹幕拖影问题报告

### DISCUSSION THREAD

**Discussion by Think0508** [Tony15246/uosc_danmaku#43] [General]
在源文件帧数较低时，弹幕会拖影。
这可以确定与源文件帧数相关。
在源bd文件《孤独摇滚》（24帧）上测试，观察到明显拖影；补帧到48帧，拖影明显改善。
24帧的《哈利·波特》同样可以复现。
得出结论：弹幕拖影与帧数有关。
这会导致大多数电影（24帧标准）受到影响。
参考：https://bbs.acgrip.com/thread-9533-1-1.html
此网址似乎找到了解决方案。

**Comment by Think0508** [comment]
但网址中的解决方案代码不准确。
可在mpv.conf中调节vf滤镜：
增加
`vf = 'fps=fps=60/1.001'`
fps后的参数参考显示器的支持帧率。
但这据信会让补帧失效。
我已成功，请试试是否可行，及是否有更好方案。
参考：https://github.com/hooke007/MPV_lazy/discussions/120

**Comment by Tony15246** [comment]
我个人自用的配置就是使用vf滤镜进行调节的，在我了解范围内暂时没听说过更优的解决方案。而且由于我穷买不起显卡😂，个人对各种视频滤镜，补帧补画质插件几乎没有使用，了解也非常有限。如果有了解的大佬可以看看有没有更好方案，提出一些建议。

**Comment by dyphire** [comment]
将脚本选项 scrolltime 的值改为 14 或 15，配合 mpv 选项 --video-sync=display-resample 可以改善拖影问题
我曾考虑将弹幕改为使用 osd-overlay 进行渲染，配合 --video-sync=display-resample 可以解决弹幕平滑度问题，但暂无法适配 xml 弹幕场景

PS：使用 vf 滤镜不是一个好主意，开销很大。只为了渲染弹幕有点过了

**Reply by Think0508** [reply]
> 将脚本选项 scrolltime 的值改为 14 或 15，配合 mpv 选项 --video-sync=display-resample 可以改善拖影问题 我曾考虑将弹幕改为使用 osd-overlay 进行渲染，配合 --video-sync=display-resample 可以解决弹幕平滑度问题，但暂无法适配 xml 弹幕场景
> 
> PS：使用 vf 滤镜不是一个好主意，开销很大。只为了渲染弹幕有点过了

试了一下，改速度对眼睛的舒适角度不是很实用……特别是在大尺寸的屏幕上更为明显。

**Reply by 1296018244** [reply]
是把sub-scrolltime=15添加到mpv.conf中去不

**Reply by Tony15246** [reply]
> 是把sub-scrolltime=15添加到mpv.conf中去不

不是的，是说在`uosc_danmaku.conf`里添加scrolltime=15。但其实不用特意设置，插件现在默认的scrolltime值就是15

**Reply by 1296018244** [reply]
哦哦，好的

**Reply by Think0508** [reply]
> 哦哦，好的

而且，这个修改影响的是弹幕的飞行速度，在屏幕上飞过的速度加快了。但刷新率还是跟着影片走的，没有变化，该24帧还是24帧，视觉效果个人觉得不太理想。

**Reply by Tony15246** [reply]
> > 哦哦，好的
> 
> 而且，这个修改影响的是弹幕的飞行速度，在屏幕上飞过的速度加快了。但刷新率还是跟着影片走的，没有变化，该24帧还是24帧，视觉效果个人觉得不太理想。

最新版本的代码重构了弹幕渲染逻辑，也做了弹幕帧率与显示帧率的适配，可以试试看有没有改善

**Reply by Think0508** [reply]
> > > 哦哦，好的
> > 
> > 
> > 而且，这个修改影响的是弹幕的飞行速度，在屏幕上飞过的速度加快了。但刷新率还是跟着影片走的，没有变化，该24帧还是24帧，视觉效果个人觉得不太理想。
> 
> 最新版本的代码重构了弹幕渲染逻辑，也做了弹幕帧率与显示帧率的适配，可以试试看有没有改善

实测，在不开启vf滤镜的前提下，还是会有肉眼可见的明显拖影（指弹幕的掉帧感）……这个恐怕只能用vf滤镜来解决了

**Reply by Tony15246** [reply]
> 实测，在不开启vf滤镜的前提下，还是会有肉眼可见的明显拖影（指弹幕的掉帧感）……这个恐怕只能用vf滤镜来解决了

看来目前只好这样了。实测过分提高弹幕帧率，性能开销甚至超过vf滤镜，而且还不一定能很好解决问题

**Reply by dyphire** [reply]
mpv 的 OSD 和字幕渲染的限制，受制于视频帧率，想要更平滑的体验只能上 vf 滤镜。或者改用其他专业的弹幕播放器，比如 kikoplay 和弹弹play。哦不对，kikoplay 也是基于 libmpv 内核，弹幕也是通过类似 OSD 绘图实现的，也要靠 vf 滤镜实现平滑效果（
